File: src/test/java/com/agiledeveloper/Fibonacci.java
package com.agiledeveloper;

import java.util.List;

import io.vavr.Tuple;
import io.vavr.collection.Stream;

class Fibonacci {
  public static int numberAtPosition(int position) {
    
    return Stream.iterate(Tuple.of(0, 1), t -> Tuple.of(t._2, t._1 + t._2))
      .drop(position)
      .get()
      ._2;
    
  }             
  
  public static List<Integer> seriesUntilPosition(int position) {
    return Stream.range(0, position + 1)
      .map(Fibonacci::numberAtPosition)
      .toJavaList();
  }
}

File: src/test/scala/com/agiledeveloper/Fibonacci.scala
package com.agiledeveloper

object Fibonacci {
  def numberAtPosition(position: Int): Int = 
    (Stream.range(0, position)
      .foldRight((0, 1): (Int, Int)) {
        case (_, (curr, next)) => (next, curr + next)
      })._2

  def seriesUntilPosition(position: Int): List[Int] =
    (0 to position).map(numberAtPosition).toList

}


